File: Python/Algorithms/LeftSumArray.py
# Given, an array of size n. Find an element that divides the array into two sub-arrays with equal sum.

def naive_method(arr):
    # slower method
    # O(n^2)
    n = len(arr)

    counter = 0
    for x in range(n):
        # calculating sum of left and right part and comparing values
        if sum(arr[:x]) == sum(arr[x + 1:]):
            counter += 1

    return counter


def fast_method(arr):
    # faster method
    # O(n)

    left_sum_array = []
    right_sum_array = []

    temp_sum = 0
    for x in arr:
        temp_sum += x
        left_sum_array.append(temp_sum)

    temp_sum = 0
    for x in reversed(arr):
        temp_sum += x
        right_sum_array.append(temp_sum)

    counter = 0
    # comparing left_sum_array and right_sum_array
    # incrementing counter whenever the values collide
    for x, y in zip(left_sum_array, right_sum_array):
        if x == y:
            counter += 1

    return counter


# Test Code

arr = [2, 3, 4, 1, 5, 4]
# Output : 1
# Subarrays are : {2, 3, 4} and {4, 5}

print(fast_method(arr))

File: CPP/Algorithms/LeftSumArray.cpp
// Given, an array of size n. Find an element that divides the array into two sub-arrays with equal sum.

#include <iostream>
#include <vector>
#include <numeric>

using namespace std;

int naive_method(vector<int> list, int n)
{
    // Slower method
    // O(n^2)
    vector<int>::iterator it;
    int counter = 0;
    for (it = list.begin(); it != list.end(); it++)
    {
        int sum1 = accumulate(list.begin(), it, 0);
        int sum2 = accumulate(it + 1, list.end(), 0);

        if (sum1 == sum2)
            counter++;
    }
    return counter;
}

int fast_method(vector<int> list, int n)
{
    // Faster method
    // O(n)

    vector<int> left_sum, right_sum;
    vector<int>::iterator it, it2;

    int counter = 0, temp_sum = 0;

    for (it = list.begin(); it != list.end(); it++)
    {
        temp_sum += *it;
        left_sum.push_back(temp_sum);
    }
    temp_sum = 0;

    for (it = list.end() - 1; it != list.begin() - 1; it--)
    {
        temp_sum += *it;
        right_sum.push_back(temp_sum);
    }

    for (it = left_sum.begin(), it2 = right_sum.begin(); it != left_sum.end(); it++, it2++)
    {
        if (*it == *it2)
            counter++;
    }

    return counter;
}

int main()
{
    // Test Code

    int n = 6;
    vector<int> vect{2, 3, 4, 1, 5, 4};
    cout << fast_method(vect, n);
}